rust/kernel/driver.rs

```diff
@@ -0,0 +1,205 @@
+// share references to it with multiple threads as nothing can be done.
+unsafe impl<T: DriverOps> Sync for Registration<T> {}
+
+impl<T: DriverOps> Registration<T> {
+    /// Creates a new instance of the registration object.
+    pub fn new() -> Self {
+        Self {
+            is_registered: false,
+            concrete_reg: UnsafeCell::new(T::RegType::default()),
+            id_table: Vec::new(),
+        }
+    }
+
+    /// Allocates a pinned registration object and registers it.
+    ///
+    /// Returns a pinned heap-allocated representation of the registration.
+    pub fn new_pinned(name: &'static CStr) -> Result<Pin<Box<Self>>> {
+        let mut reg = Pin::from(Box::try_new(Self::new())?);
+        reg.as_mut().register(name)?;
+        Ok(reg)
+    }
+
+    /// Registers a driver with its subsystem.
+    ///
+    /// It must be pinned because the memory block that represents the registration is potentially
+    /// self-referential.
+    pub fn register(self: Pin<&mut Self>, name: &'static CStr) -> Result {
+        // SAFETY: We never move out of `this`.
+        let this = unsafe { self.get_unchecked_mut() };
+        if this.is_registered {
+            // Already registered.
+            return Err(Error::EINVAL);
+        }
+
+        if this.id_table.is_empty() {
+            this.build_table()?;
+        }
+
+        // SAFETY: `concrete_reg` was initialised via its default constructor. `id_table` was just
+        // initialised above with a zero terminating entry. Both are only freed after `Self::drop`
+        // is called, which first calls `T::unregister`.
+        unsafe {
+            T::register(
+                this.concrete_reg.get(),
+                name,
+                &this.id_table[0] as *const _ as *const _,
+            )
+        }?;
+
+        this.is_registered = true;
+        Ok(())
+    }
+
+    /// Builds the zero-terminated raw-type array of supported devices.
+    ///
+    /// This is not ideal because the table is built at runtime. Once Rust fully supports const
+    /// generics, we can build the table at compile time.
+    fn build_table(&mut self) -> Result {
+        // Clear the table on failure, to indicate that the table isn't initialised.
+        let mut table = ScopeGuard::new_with_data(&mut self.id_table, |t| t.clear());
+
+        table.try_reserve_exact(T::ID_TABLE.len() + 1)?;
+        for (i, id) in T::ID_TABLE.iter().enumerate() {
+            table.try_push(MaybeUninit::new(T::to_raw_id(i, id)))?;
+        }
+        table.try_push(MaybeUninit::uninit())?;
+
+        // SAFETY: There is always a last entry on the table (because of the insertion at the line
+        // above), and it is valid for write and properly aligned. Even if a zeroed version is an
+        // invalid representation of `T::RawIdType`, it is never interpreted as such because it's a
+        // `MaybeUninit`.
+        unsafe { core::intrinsics::write_bytes(table.last_mut().unwrap().as_mut_ptr(), 0, 1) };
```

Comment:
Can you push `MaybeUninit::zeroed()`? The current code risks accidentally overwriting the last zero due to not extending the length. Also I think this only writes a single byte and not a full zero entry.

Reply:
I don't follow what you mean by 'accidentally overwriting the last zero due to not extending the length'? If we can't extend the length, the function will fail.

Regarding writing a single byte, the [documentation](https://doc.rust-lang.org/std/ptr/fn.write_bytes.html) says that `count * size_of::<T>()` bytes are written, so unless this is a mistake (which should then be fixed), this is writing over the whole `T`.

All of this said, `MaybeUninit::zeroed` is definitely better, with the added benefit of one less unsafe block. I've changed the code to use it, thanks!

Reply:
> I don't follow what you mean by 'accidentally overwriting the last zero due to not extending the length'? If we can't extend the length, the function will fail.

Right, missed the `table.try_push(MaybeUninit::uninit())?;` before the `write_bytes`.

> Regarding writing a single byte, the documentation says that count * size_of::<T>() bytes are written, so unless this is a mistake (which should then be fixed), this is writing over the whole T.

:+1: